event/bus.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package event
+
+import (
+	"sync/atomic"
+)
+
+type subscribeCommand *Subscription
+
+type unsubscribeCommand *Subscription
+
+type publishCommand Message
+
+type closeCommand struct{}
+
+// Message contains event info.
+type Message struct {
+	// Name is the name of the event this message was generated from.
+	Name string
+	// Data contains optional event information.
+	Data any
+}
+
+// NewMessage returns a new message with the given name and optional data.
+func NewMessage(name string, data any) Message {
+	return Message{name, data}
+}
+
+// Subscription is a read-only event stream.
+type Subscription struct {
+	id     uint64
+	value  chan Message
+	events []string
+}
+
+// Message returns the next event value from the subscription.
+func (s *Subscription) Message() <-chan Message {
+	return s.value
+}
+
+// Events returns the names of all subscribed events.
+func (s *Subscription) Events() []string {
+	return s.events
+}
+
+// Bus is used to broadcast events to subscribers.
+type Bus struct {
```

Comment:
todo: I thought we agreed to keep this behind an interface?  Happy to resume that conversation if need be :)

Reply:
It already exists in the `events` package, this PR deleted it.  I think it would just need a rename, removal of the generic, and adding the `...string` param to `Subscribe`, otherwise it looks identical to the `*Bus` signature.

Reply:
It doesn't make sense to define the interface in the `event` package has it doesn't need/use it. We could define it in  `client` though for the purpose of the `client.DB` interface method `Events() Events`.

Reply:
If we are going to move it, I'd also suggest moving `Subscription` there too, and then the `event` package can become internal.

Otherwise, I do slightly disagree with you.  It is quite common for packages to define their public interfaces within themselves.  Especially given `event` is not really Defra specific, and could quite happily be moved outside of this repo for general use.

Reply:
Packages defining public interfaces that are not used internally is bad practice though. We should try to avoid it. The goal of an interface is to make something more flexible. Like if we want `db` to be able to use different event bus implementation. I'd argue that we are building the `Bus` specifically for our needs so I don't think it flexibility makes sense here. I much prefer the concrete type as it is now.

Reply:
I added the interface because I will be using it to create a JavaScript event bus for the WASM build.